pkg/controller/direct/compute/firewallpolicyrule/refs.go

```diff
@@ -0,0 +1,212 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package firewallpolicyrule
+
+import (
+	"context"
+	"fmt"
+
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/compute/v1beta1"
+
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/k8s"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func ResolveComputeFirewallPolicy(ctx context.Context, reader client.Reader, src client.Object, ref *refs.ComputeFirewallPolicyRef) (*refs.ComputeFirewallPolicyRef, error) {
+	if ref == nil {
+		return nil, nil
+	}
+
+	if ref.External != "" {
+		if ref.Name != "" {
+			return nil, fmt.Errorf("cannot specify both name and external on reference")
+		}
+		return ref, nil
+	}
+
+	if ref.Name == "" {
+		return nil, fmt.Errorf("must specify either name or external on reference")
+	}
+
+	key := types.NamespacedName{
+		Namespace: ref.Namespace,
+		Name:      ref.Name,
+	}
+	if key.Namespace == "" {
+		key.Namespace = src.GetNamespace()
+	}
+
+	computeFirwallPolicy, err := resolveResourceName(ctx, reader, key, schema.GroupVersionKind{
+		Group:   "compute.cnrm.cloud.google.com",
+		Version: "v1beta1",
+		Kind:    "ComputeFirewallPolicy",
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	resourceID, err := refs.GetResourceID(computeFirwallPolicy)
+	if err != nil {
+		return nil, err
+	}
+
+	return &refs.ComputeFirewallPolicyRef{
+		External: fmt.Sprintf("%s", resourceID)}, nil
+}
+
+func ResolveComputeNetwork(ctx context.Context, reader client.Reader, src client.Object, ref *refs.ComputeNetworkRef) (*refs.ComputeNetworkRef, error) {
+	if ref == nil {
+		return nil, nil
+	}
+
+	if ref.External != "" {
+		if ref.Name != "" {
+			return nil, fmt.Errorf("cannot specify both name and external on reference")
+		}
+		return ref, nil
+	}
+
+	if ref.Name == "" {
+		return nil, fmt.Errorf("must specify either name or external on reference")
+	}
+
+	key := types.NamespacedName{
+		Namespace: ref.Namespace,
+		Name:      ref.Name,
+	}
+	if key.Namespace == "" {
+		key.Namespace = src.GetNamespace()
+	}
+
+	computeNetwork, err := resolveResourceName(ctx, reader, key, schema.GroupVersionKind{
+		Group:   "compute.cnrm.cloud.google.com",
+		Version: "v1beta1",
+		Kind:    "ComputeNetwork",
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	resourceID, err := refs.GetResourceID(computeNetwork)
+	if err != nil {
+		return nil, err
+	}
+
+	projectID, err := refs.ResolveProjectID(ctx, reader, computeNetwork)
+	if err != nil {
+		return nil, err
+	}
+
+	return &refs.ComputeNetworkRef{
+		External: fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%s/global/networks/%s", projectID, resourceID)}, nil
```

Comment:
Nit: I would have expected `projects/%s/...` (no "https://...").  We should try to match the format that we want users to use.

Reply:
Reformat the value because we want to match the realGCP log. In realGCP log, it looks like:
```
  "targetResources": [
    "https://www.googleapis.com/compute/v1/projects/${projectId}/global/networks/network-${uniqueId}"
  ],
```
Also the target resource url requires the prefix, if I remove it, the creation will fail with error: `Invalid target resource url: projects/cnrm-yuhou/global/networks/network-bflfo2mcesdexjq`